### ISSUE TITLE
Old ruby (2.0.0) cannot read gzip files created by samtools

### ISSUE THREAD

**Issue by fplaza** [samtools/samtools#1831]
Hi,

I have updated samtools from version 1.9 to version 1.17
I have noticed that reading a gzip compressed fastq file generated by samtools 1.17  fails with a ruby script run with ruby 2.0.0 (installed by default in centos 7). More precisely, the reading ends prematurely. 
This issue never occured with samtools 1.9.

After investigation, it seems to be a bug in ruby (https://bugs.ruby-lang.org/issues/9790), not in samtools.
 However, I was wondering what changed in the way samtools writes gzip files between version 1.9 to version 1.17?

Thanks,
Florian

**Comment by fplaza** [comment]
Both are linked to libz.so according to ldd
The command you give returns nothing  for both executables

**Comment by jkbonfield** [comment]
Hmm, not sure then.

There have been numerous little bug fixes to bgzf over the years, and bgzip also gained a text mode that attempts to detect line breaks and keep records within a single bgzf block, but I can't really say why any of those would break Ruby.

Either way, the solution really should be to fix Ruby!  I assume "zcat" is happy to decompress your data?

Also I would *strongly* advise that you build htslib with libdeflate.  It's far more performant than zlib.

**Comment by fplaza** [comment]
I am using a combination of bowtie2 + samtools to filter out reads from one fastq file mapped to a reference genome.
Actually, I don't understand why the new samtools concatenates multiple streams in the final gzipped file.
Is it because of parallelisation?

Of course, we need to update ruby but I was curious to understand the reason behind all of this.
Regarding libdeflate, I am not sure it is available on centos 7.

**Comment by jkbonfield** [comment]
Concatenation of multiple streams is just how the BGZF format works, but yes it's to do with parallelism.  It's been that way forever, so I really don't see how 1.9 worked unless the fastq.gz output isn't actually bgzfed and is a single naive gzip stream (and hence not parallelisable).  We did rewrite a lot of samtools fastq, so it's possible, but if you want efficiency then that basically means BGZF and sadly problems for old Ruby.  The alternative is output to uncompressed fastq instead (and maybe use pigz afterwards, but that may also be using concatenated gz streams).

Closing this because as you say the solution is basically to upgrade Ruby (or Centos).
